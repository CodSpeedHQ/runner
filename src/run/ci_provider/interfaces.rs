use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::BTreeMap;

#[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
#[serde(rename_all = "UPPERCASE")]
pub enum RepositoryProvider {
    GitLab,
    GitHub,
}

#[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum Platform {
    GithubActions,
    GitlabCi,
    Buildkite,
    Local,
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct CIProviderMetadata {
    #[serde(rename = "ref")]
    pub ref_: String,
    pub head_ref: Option<String>,
    pub base_ref: Option<String>,
    pub owner: String,
    pub repository: String,
    pub event: RunEvent,
    pub sender: Option<Sender>,
    pub gh_data: Option<GhData>,
    pub gl_data: Option<GlData>,
    pub repository_root_path: String,
}

#[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum RunEvent {
    Push,
    PullRequest,
    WorkflowDispatch,
    Schedule,
    Local,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GhData {
    pub run_id: String,
    pub job: String,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GlData {
    pub run_id: String,
    pub job: String,
}

/// Each execution of the CLI maps to a `RunPart`.
///
/// Several `RunParts` can be aggregated in a single `Run` thanks to this data.
#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct RunPart {
    /// A unique identifier of the `Run` on the platform
    pub run_id: String,

    /// Uniquely identify a `RunPart` within a `Run`.
    ///
    /// This id can be the same between `RunParts` of different `Runs`.
    pub run_part_id: String,

    /// The name of the job. For example, on Github Actions, the workflow name.
    ///
    /// This is **not** unique between executions of the CLI, even between matrix jobs.
    pub job_name: String,

    /// Some relevant metadata.
    ///
    /// This can include matrix and strategy for GithubActions,
    /// some relevant env values.
    ///
    /// We use a `BTreeMap` and not a `HashMap` to keep insert order for
    /// `serde_json` serialization.
    pub metadata: BTreeMap<String, Value>,
}

#[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct Sender {
    pub id: String,
    pub login: String,
}
