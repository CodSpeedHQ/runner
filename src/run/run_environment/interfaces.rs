use serde::{Deserialize, Serialize};

#[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
#[serde(rename_all = "UPPERCASE")]
pub enum RepositoryProvider {
    GitLab,
    GitHub,
}

#[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum RunEnvironment {
    GithubActions,
    GitlabCi,
    Buildkite,
    Local,
}

#[derive(Deserialize, Serialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct RunEnvironmentMetadata {
    #[serde(rename = "ref")]
    pub ref_: String,
    pub head_ref: Option<String>,
    pub base_ref: Option<String>,
    pub owner: String,
    pub repository: String,
    pub event: RunEvent,
    pub sender: Option<Sender>,
    pub gh_data: Option<GhData>,
    pub gl_data: Option<GlData>,
    pub repository_root_path: String,
}

#[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum RunEvent {
    Push,
    PullRequest,
    WorkflowDispatch,
    Schedule,
    Local,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GhData {
    pub run_id: String,
    pub job: String,
}

#[derive(Deserialize, Serialize, Debug, Clone)]
#[serde(rename_all = "camelCase")]
pub struct GlData {
    pub run_id: String,
    pub job: String,
}

#[derive(Deserialize, Serialize, Debug, Clone, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct Sender {
    pub id: String,
    pub login: String,
}
