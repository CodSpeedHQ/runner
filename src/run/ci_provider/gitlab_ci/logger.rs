use console::style;
use lazy_static::lazy_static;
use log::{Level, LevelFilter, Log};
use regex::Regex;
use simplelog::SharedLogger;
use std::{
    env,
    io::Write,
    sync::Mutex,
    time::{SystemTime, UNIX_EPOCH},
};

use crate::{
    logger::{get_group_event, GroupEvent},
    run::ci_provider::logger::should_provider_logger_handle_record,
};

lazy_static! {
    static ref GITLAB_SECTION_ID_SANITIZE_REGEX: Regex =
        Regex::new(r"[^\d\w\-_]").expect("Failed to compile GitLab SectionId regex");
}

/// Unicode Escape character
///
/// https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#general-ascii-codes
const U_ESC: char = '\x1B';

/// Unicode Carriage Return character
///
/// https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#general-ascii-codes
const U_CR: char = '\x0D';

/// Reset color mode
///
/// https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#colors--graphics-mode
const COLOR_RESET: &str = "\x1B[0m";

/// Erase cursor till end of line
///
/// https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#erase-functions
const ERASE_CURSOR: &str = "\x1B[0K";

/// A logger that prints log in the format expected by GitLab CI
///
/// See https://docs.gitlab.com/ee/ci/yaml/script.html
pub struct GitLabCILogger {
    log_level: LevelFilter,
    section_id: Mutex<Option<String>>,
}

impl GitLabCILogger {
    pub fn new() -> Self {
        // force activation of colors, because GitlabCI does not
        // respect the CLICOLORS spec.
        // https://gitlab.com/gitlab-org/gitlab/-/issues/28598
        console::set_colors_enabled(true);

        let log_level = env::var("CODSPEED_LOG")
            .ok()
            .and_then(|log_level| log_level.parse::<log::LevelFilter>().ok())
            .unwrap_or(log::LevelFilter::Info);
        Self {
            log_level,
            section_id: Mutex::new(None),
        }
    }
}

impl Log for GitLabCILogger {
    fn enabled(&self, _metadata: &log::Metadata) -> bool {
        true
    }

    fn log(&self, record: &log::Record) {
        if !should_provider_logger_handle_record(record) {
            return;
        }

        let level = record.level();
        let message = record.args();

        if let Some(group_event) = get_group_event(record) {
            let now = SystemTime::now();
            let timestamp = now.duration_since(UNIX_EPOCH).unwrap().as_secs();
            let mut section_id = self.section_id.lock().unwrap();

            match group_event {
                GroupEvent::Start(name) | GroupEvent::StartOpened(name) => {
                    let new_section_id = GITLAB_SECTION_ID_SANITIZE_REGEX
                        .replace_all(&name, "_")
                        .to_ascii_lowercase();

                    *section_id = Some(new_section_id.to_string());

                    // https://docs.gitlab.com/ee/ci/yaml/script.html#custom-collapsible-sections
                    println!("{ERASE_CURSOR}section_start:{timestamp}:{new_section_id}{U_CR}{ERASE_CURSOR}{U_ESC}[36;1m{name}{COLOR_RESET}");
                }
                GroupEvent::End => {
                    // do not fail if there is no current section
                    let current_section_id = section_id.clone().unwrap_or("".to_string());

                    // https://docs.gitlab.com/ee/ci/yaml/script.html#custom-collapsible-sections
                    println!("{ERASE_CURSOR}section_end:{timestamp}:{current_section_id}{U_CR}{ERASE_CURSOR}");

                    *section_id = None;
                }
            }
            return;
        }

        if level > self.log_level {
            return;
        }

        // set log colors. See https://gist.github.com/fnky/458719343aabd01cfb17a3a4f7296797#colors--graphics-mode
        match level {
            Level::Error => {
                println!("{}", style(message).red());
            }
            Level::Warn => {
                println!("{}", style(message).yellow());
            }
            Level::Info => {
                println!("{message}");
            }
            Level::Debug => {
                println!("{}", style(message).cyan());
            }
            Level::Trace => {
                println!("{}", style(message).magenta());
            }
        };
    }

    fn flush(&self) {
        std::io::stdout().flush().unwrap();
    }
}

impl SharedLogger for GitLabCILogger {
    fn level(&self) -> LevelFilter {
        self.log_level
    }

    fn config(&self) -> Option<&simplelog::Config> {
        None
    }

    fn as_log(self: Box<Self>) -> Box<dyn Log> {
        Box::new(*self)
    }
}
