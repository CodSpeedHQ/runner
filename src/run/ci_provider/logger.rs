use crate::run::runner::VALGRIND_EXECUTION_TARGET;

pub(super) fn should_provider_logger_handle_record(record: &log::Record) -> bool {
    // Provider logger should handle all records except the ones from the valgrind execution target
    record.target() != VALGRIND_EXECUTION_TARGET
}
