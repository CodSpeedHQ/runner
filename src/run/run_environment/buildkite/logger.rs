use crate::{
    logger::{get_group_event, get_json_event, GroupEvent},
    run::run_environment::logger::should_provider_logger_handle_record,
};
use log::*;
use simplelog::SharedLogger;
use std::{env, io::Write};

/// A logger that prints logs in the format expected by Buildkite
///
/// See https://buildkite.com/docs/pipelines/managing-log-output
pub struct BuildkiteLogger {
    log_level: LevelFilter,
}

impl BuildkiteLogger {
    pub fn new() -> Self {
        let log_level = env::var("CODSPEED_LOG")
            .ok()
            .and_then(|log_level| log_level.parse::<log::LevelFilter>().ok())
            .unwrap_or(log::LevelFilter::Info);
        Self { log_level }
    }
}

impl Log for BuildkiteLogger {
    fn enabled(&self, _metadata: &Metadata) -> bool {
        true
    }

    fn log(&self, record: &Record) {
        if !should_provider_logger_handle_record(record) {
            return;
        }

        let level = record.level();
        let message = record.args();

        if let Some(group_event) = get_group_event(record) {
            match group_event {
                GroupEvent::Start(name) => {
                    println!("--- {name}");
                }
                GroupEvent::StartOpened(name) => {
                    println!("+++ {name}");
                }
                GroupEvent::End => {}
            }
            return;
        }

        if get_json_event(record).is_some() {
            return;
        }

        if level > self.log_level {
            return;
        }
        // there is no support for log levels in Buildkite, so we print the level in the message
        match level {
            Level::Error => {
                println!("[ERROR] {message}");
            }
            Level::Warn => {
                println!("[WARN] {message}");
            }
            Level::Info => {
                println!("{message}");
            }
            Level::Debug => {
                println!("[DEBUG] {message}");
            }
            Level::Trace => {
                println!("[TRACE] {message}");
            }
        }
    }

    fn flush(&self) {
        std::io::stdout().flush().unwrap();
    }
}

impl SharedLogger for BuildkiteLogger {
    fn level(&self) -> LevelFilter {
        self.log_level
    }

    fn config(&self) -> Option<&simplelog::Config> {
        None
    }

    fn as_log(self: Box<Self>) -> Box<dyn Log> {
        Box::new(*self)
    }
}
