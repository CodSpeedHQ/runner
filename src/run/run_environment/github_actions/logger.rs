use crate::{
    logger::{GroupEvent, get_group_event, get_json_event},
    run::run_environment::logger::should_provider_logger_handle_record,
};
use log::*;
use simplelog::SharedLogger;
use std::{env, io::Write};

/// A logger that prints logs in the format expected by GitHub Actions, with grouping support.
///
/// See https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions
pub struct GithubActionLogger {
    log_level: LevelFilter,
}

impl GithubActionLogger {
    pub fn new() -> Self {
        // Only enable debug logging if it's enabled in GitHub Actions.
        // See: https://docs.github.com/en/actions/reference/workflows-and-actions/variables
        let log_level = if env::var("RUNNER_DEBUG").unwrap_or_default() == "1" {
            LevelFilter::Trace
        } else {
            env::var("CODSPEED_LOG")
                .ok()
                .and_then(|log_level| log_level.parse::<LevelFilter>().ok())
                .unwrap_or(LevelFilter::Info)
        };

        Self { log_level }
    }
}

impl Log for GithubActionLogger {
    fn enabled(&self, _metadata: &Metadata) -> bool {
        true
    }

    fn log(&self, record: &Record) {
        if !should_provider_logger_handle_record(record) {
            return;
        }

        let level = record.level();
        let message = record.args();

        if let Some(group_event) = get_group_event(record) {
            match group_event {
                GroupEvent::Start(name) | GroupEvent::StartOpened(name) => {
                    println!("::group::{name}");
                }
                GroupEvent::End => {
                    println!("::endgroup::");
                }
            }
            return;
        }

        if get_json_event(record).is_some() {
            return;
        }

        if level > self.log_level {
            return;
        }

        let prefix = match level {
            Level::Error => "::error::",
            Level::Warn => "::warning::",
            Level::Info => "",
            Level::Debug => "::debug::",
            Level::Trace => "::debug::[TRACE]",
        };
        let message_string = message.to_string();
        let lines = message_string.lines();
        // ensure that all the lines of the message have the prefix, otherwise GitHub Actions will not recognize the command for the whole string
        lines.for_each(|line| println!("{prefix}{line}"));
    }

    fn flush(&self) {
        std::io::stdout().flush().unwrap();
    }
}

impl SharedLogger for GithubActionLogger {
    fn level(&self) -> LevelFilter {
        self.log_level
    }

    fn config(&self) -> Option<&simplelog::Config> {
        None
    }

    fn as_log(self: Box<Self>) -> Box<dyn Log> {
        Box::new(*self)
    }
}
