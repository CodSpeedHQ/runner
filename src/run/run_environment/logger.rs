use crate::run::runner::EXECUTOR_TARGET;

pub(super) fn should_provider_logger_handle_record(record: &log::Record) -> bool {
    // Provider logger should handle all records except the ones from the executor target
    record.target() != EXECUTOR_TARGET
}
