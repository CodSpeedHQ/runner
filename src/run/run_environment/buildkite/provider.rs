use std::env;

use simplelog::SharedLogger;

use crate::prelude::*;
use crate::run::helpers::{parse_git_remote, GitRemote};
use crate::run::run_environment::interfaces::RunEnvironment;
use crate::run::{
    config::Config,
    helpers::{find_repository_root, get_env_variable},
    run_environment::{
        interfaces::{RepositoryProvider, RunEnvironmentMetadata, RunEvent},
        provider::{RunEnvironmentDetector, RunEnvironmentProvider},
    },
};

use super::logger::BuildkiteLogger;

#[derive(Debug)]
pub struct BuildkiteProvider {
    owner: String,
    repository: String,
    ref_: String,
    head_ref: Option<String>,
    base_ref: Option<String>,
    event: RunEvent,
    repository_root_path: String,
}

pub fn get_pr_number() -> Result<Option<u64>> {
    Ok(get_env_variable("BUILDKITE_PULL_REQUEST")?.parse().ok())
}

pub fn get_run_event() -> Result<RunEvent> {
    let is_pr = get_pr_number()?.is_some();

    if is_pr {
        Ok(RunEvent::PullRequest)
    } else {
        Ok(RunEvent::Push)
    }
}

pub fn get_ref() -> Result<String> {
    let pr_number = get_pr_number()?;

    if let Some(pr_number) = pr_number {
        Ok(format!("refs/pull/{}/merge", pr_number))
    } else {
        Ok(format!(
            "refs/heads/{}",
            get_env_variable("BUILDKITE_BRANCH")?
        ))
    }
}

impl TryFrom<&Config> for BuildkiteProvider {
    type Error = Error;
    fn try_from(config: &Config) -> Result<Self> {
        if config.token.is_none() {
            bail!("Token authentication is required for Buildkite");
        }

        let is_pr = get_pr_number()?.is_some();
        let repository_url = get_env_variable("BUILDKITE_REPO")?;
        let GitRemote {
            owner,
            repository,
            domain,
        } = parse_git_remote(&repository_url)?;

        if domain != "github.com" {
            bail!(
                "Only GitHub repositories are supported by CodSpeed BuildKite integration for now."
            );
        }

        let repository_root_path = match find_repository_root(&std::env::current_dir()?) {
            Some(mut path) => {
                // Add a trailing slash to the path
                path.push("");
                path.to_string_lossy().to_string()
            }
            None => format!(
                "/buildkite/builds/{}/{}/{}/",
                get_env_variable("BUILDKITE_AGENT_NAME")?,
                get_env_variable("BUILDKITE_ORGANIZATION_SLUG")?,
                get_env_variable("BUILDKITE_PIPELINE_SLUG")?,
            ),
        };

        Ok(Self {
            owner: owner.clone(),
            repository: repository.clone(),
            ref_: get_ref()?,
            base_ref: if is_pr {
                Some(get_env_variable("BUILDKITE_PULL_REQUEST_BASE_BRANCH")?)
            } else {
                None
            },
            head_ref: if is_pr {
                Some(get_env_variable("BUILDKITE_BRANCH")?)
            } else {
                None
            },
            event: get_run_event()?,
            repository_root_path,
        })
    }
}

impl RunEnvironmentDetector for BuildkiteProvider {
    fn detect() -> bool {
        env::var("BUILDKITE") == Ok("true".into())
    }
}

impl RunEnvironmentProvider for BuildkiteProvider {
    fn get_repository_provider(&self) -> RepositoryProvider {
        RepositoryProvider::GitHub
    }

    fn get_logger(&self) -> Box<dyn SharedLogger> {
        Box::new(BuildkiteLogger::new())
    }

    fn get_run_environment_name(&self) -> &'static str {
        "Buildkite"
    }

    fn get_run_environment(&self) -> RunEnvironment {
        RunEnvironment::Buildkite
    }

    fn get_run_environment_metadata(&self) -> Result<RunEnvironmentMetadata> {
        Ok(RunEnvironmentMetadata {
            base_ref: self.base_ref.clone(),
            head_ref: self.head_ref.clone(),
            event: self.event.clone(),
            owner: self.owner.clone(),
            repository: self.repository.clone(),
            ref_: self.ref_.clone(),
            repository_root_path: self.repository_root_path.clone(),
            gh_data: None,
            gl_data: None,
            sender: None,
        })
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_json_snapshot;
    use temp_env::{with_var, with_vars};

    use crate::VERSION;

    use super::*;

    #[test]
    fn test_detect() {
        with_var("BUILDKITE", Some("true"), || {
            assert!(BuildkiteProvider::detect());
        });
    }

    #[test]
    fn test_try_from_push_main() {
        with_vars(
            [
                ("BUILDKITE_AGENT_NAME", Some("7b10eca7600b-1")),
                ("BUILDKITE_BRANCH", Some("main")),
                ("BUILDKITE_BUILD_NUMBER", Some("1")),
                ("BUILDKITE_COMMIT", Some("abc123")),
                ("BUILDKITE_ORGANIZATION_SLUG", Some("my-org")),
                ("BUILDKITE_PIPELINE_SLUG", Some("buildkite-test")),
                ("BUILDKITE_PULL_REQUEST_BASE_BRANCH", Some("")),
                ("BUILDKITE_PULL_REQUEST", Some("")),
                (
                    "BUILDKITE_REPO",
                    Some("https://github.com/my-org/adrien-python-test.git"),
                ),
                ("BUILDKITE", Some("true")),
            ],
            || {
                let config = Config {
                    token: Some("token".into()),
                    ..Config::test()
                };
                let provider = BuildkiteProvider::try_from(&config).unwrap();

                assert_eq!(provider.owner, "my-org");
                assert_eq!(provider.repository, "adrien-python-test");
                assert_eq!(provider.ref_, "refs/heads/main");
                assert_eq!(provider.base_ref, None);
                assert_eq!(provider.head_ref, None);
                assert_eq!(provider.event, RunEvent::Push);
                assert_eq!(
                    provider.repository_root_path,
                    "/buildkite/builds/7b10eca7600b-1/my-org/buildkite-test/"
                );
            },
        );
    }

    #[test]
    fn test_try_from_pull_request() {
        with_vars(
            [
                ("BUILDKITE_AGENT_NAME", Some("7b10eca7600b-1")),
                ("BUILDKITE_BRANCH", Some("feat/codspeed-runner")),
                ("BUILDKITE_BUILD_NUMBER", Some("1")),
                ("BUILDKITE_COMMIT", Some("abc123")),
                ("BUILDKITE_ORGANIZATION_SLUG", Some("my-org")),
                ("BUILDKITE_PIPELINE_SLUG", Some("buildkite-test")),
                ("BUILDKITE_PULL_REQUEST_BASE_BRANCH", Some("main")),
                ("BUILDKITE_PULL_REQUEST", Some("22")),
                (
                    "BUILDKITE_REPO",
                    Some("git@github.com:my-org/adrien-python-test.git"),
                ),
                ("BUILDKITE", Some("true")),
            ],
            || {
                let config = Config {
                    token: Some("token".into()),
                    ..Config::test()
                };
                let provider = BuildkiteProvider::try_from(&config).unwrap();

                assert_eq!(provider.owner, "my-org");
                assert_eq!(provider.repository, "adrien-python-test");
                assert_eq!(provider.ref_, "refs/pull/22/merge");
                assert_eq!(provider.base_ref, Some("main".into()));
                assert_eq!(provider.head_ref, Some("feat/codspeed-runner".into()));
                assert_eq!(provider.event, RunEvent::PullRequest);
                assert_eq!(
                    provider.repository_root_path,
                    "/buildkite/builds/7b10eca7600b-1/my-org/buildkite-test/"
                );
            },
        );
    }

    #[test]
    fn test_pull_request_run_environment_metadata() {
        with_vars(
            [
                ("BUILDKITE_AGENT_NAME", Some("7b10eca7600b-1")),
                ("BUILDKITE_BRANCH", Some("feat/codspeed-runner")),
                ("BUILDKITE_BUILD_NUMBER", Some("1")),
                ("BUILDKITE_COMMIT", Some("abc123")),
                ("BUILDKITE_ORGANIZATION_SLUG", Some("my-org")),
                ("BUILDKITE_PIPELINE_SLUG", Some("buildkite-test")),
                ("BUILDKITE_PULL_REQUEST_BASE_BRANCH", Some("main")),
                ("BUILDKITE_PULL_REQUEST", Some("22")),
                (
                    "BUILDKITE_REPO",
                    Some("git@github.com:my-org/adrien-python-test.git"),
                ),
                ("BUILDKITE", Some("true")),
            ],
            || {
                let config = Config {
                    token: Some("token".into()),
                    ..Config::test()
                };
                let provider = BuildkiteProvider::try_from(&config).unwrap();
                let run_environment_metadata = provider.get_run_environment_metadata().unwrap();

                assert_json_snapshot!(run_environment_metadata, {
                    ".runner.version" => insta::dynamic_redaction(|value,_path| {
                        assert_eq!(value.as_str().unwrap(), VERSION.to_string());
                        "[version]"
                    }),
                })
            },
        );
    }
}
