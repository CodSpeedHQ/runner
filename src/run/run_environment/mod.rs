pub mod interfaces;
pub mod logger;
mod provider;

use buildkite::BuildkiteProvider;
use github_actions::GitHubActionsProvider;
use gitlab_ci::GitLabCIProvider;
use local::LocalProvider;
use provider::RunEnvironmentDetector;

use crate::prelude::*;
use crate::run::config::Config;

pub use self::interfaces::*;
pub use self::provider::RunEnvironmentProvider;

// RunEnvironment Provider implementations
mod buildkite;
mod github_actions;
mod gitlab_ci;
mod local;

pub fn get_provider(config: &Config) -> Result<Box<dyn RunEnvironmentProvider>> {
    if BuildkiteProvider::detect() {
        let provider = BuildkiteProvider::try_from(config)?;
        return Ok(Box::new(provider));
    }

    if GitHubActionsProvider::detect() {
        let provider = GitHubActionsProvider::try_from(config)?;
        return Ok(Box::new(provider));
    }

    if GitLabCIProvider::detect() {
        let provider = GitLabCIProvider::try_from(config)?;
        return Ok(Box::new(provider));
    }

    if LocalProvider::detect() {
        let provider = LocalProvider::try_from(config)?;
        return Ok(Box::new(provider));
    }

    // By design, this should not happen as the `LocalProvider` is a fallback
    bail!("No RunEnvironment provider detected")
}
