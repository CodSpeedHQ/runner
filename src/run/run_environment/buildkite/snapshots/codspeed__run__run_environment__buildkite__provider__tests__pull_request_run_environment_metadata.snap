---
source: src/run/run_environment/buildkite/provider.rs
expression: run_environment_metadata
---
{
  "ref": "refs/pull/22/merge",
  "headRef": "feat/codspeed-runner",
  "baseRef": "main",
  "owner": "my-org",
  "repository": "adrien-python-test",
  "event": "pull_request",
  "sender": null,
  "ghData": null,
  "glData": null,
  "repositoryRootPath": "/buildkite/builds/7b10eca7600b-1/my-org/buildkite-test/"
}
